### ISSUE TITLE
Support serializing/deserializing Python classes

### ISSUE THREAD

**Issue by justinvp** [pulumi/pulumi#4789]
Support for (un)marshaling nested inputs/outputs classes.

Part of https://github.com/pulumi/pulumi/issues/3771.

**Comment by justinvp** [comment]
This is mostly finished. Late last week, I ran into a roadblock introspecting generic type annotations at runtime on Python 3.6/3.7, so moved on to working on #4790 to unblock downstream work using a slightly more verbose syntax (actual usage is the same either way), with the intent to come back to this and workaround the roadblock, before checking this in. Planning to get back to this later today.

**Comment by justinvp** [comment]
Python 3.7 is working and I think I have a path forward for Python 3.6.

**Comment by justinvp** [comment]
Aiming to have a PR out for this today.

**Comment by justinvp** [comment]
Testing and polishing took longer than I expected, but the implementation is nearly ready to send out for review (along with provider codegen changes for #4790).

In the meantime, I'd love feedback on the syntax for declaring these. Making changes should be quick & easy at this point as I've already tried many variations (e.g. using base classes, decorating property getters, etc.).

## Input Types

Here's a simple input type with a required property and optional property:

```python
from pulumi import Input, input_property, input_type

@input_type
class FooArgs:
    first_arg: Input[str] = input_property("firstArg")
    second_arg: Optional[Input[float]] = input_property("secondArg")

    def __init__(self, first_arg: Input[str], second_arg: Optional[Input[float]]=None):
        self.first_arg = first_arg
        self.second_arg = second_arg
```

This is effectively what our provider generated code looks like (though, the _actual_ generated code has docstrings, does runtime validation of args in `__init__()`, and has a couple other minor differences).

The syntax isn't overly reliant on code generation and is pleasant enough to write by hand (e.g. when writing dynamic providers).

The `input_property()` function is inspired by [`dataclasses.field()`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field). It's used to specify the Pulumi property name so the correct name is sent to the engine.

See footnote [1] below for some questions and further thoughts on `input_property()`.

If you don't care about mapping snake_case property names to camelCase, you could drop the usage of `input_property()`:

```python
@input_type
class FooArgs:
    first_arg: Input[str]
    second_arg: Optional[Input[float]]

    def __init__(self, first_arg: Input[str], second_arg: Optional[Input[float]]=None):
        self.first_arg = first_arg
        self.second_arg = second_arg
```

And if you are using Python 3.7+, you could further simplify by making use of `dataclasses` to avoid having to define an `__init__()` at all.

```python
@dataclass
@input_type
class FooArgs:
    first_arg: Input[str]
    second_arg: Optional[Input[float]] = None
```

Of course, you can also still use `input_property` with `@dataclass`:

```python
@dataclass
@input_type
class FooArgs:
    first_arg: Input[str] = input_property("firstArg")
    second_arg: Optional[Input[float]] = input_property("secondArg", None)
```

> Note: I'm tempted to make the `input_type` decorator handle adding an `__init__()` automatically if one isn't already declared, similar to what `dataclasses.dataclass` does. And possibly also automatically add `__eq__`,  `__ne__`, `__repr__` (if not already declared). 🤔

## Output Types

Output types are similar to input types:

```python
from pulumi import output_property, output_type

@output_type
class Foo(dict):
    first_arg: str = output_property("firstArg")
    second_arg: float = output_property("secondArg")
```

This is effectively what our generated code looks like (though, the _actual_ generated code has docstrings).

Inheriting from `dict` is optional. We do it in our generated code to maintain compatibility, but you don't have to. By not subclassing `dict`, you can keep intellisense clean for your objects (avoids all of `dict`'s methods).

You also don't need to use `output_property` if you don't care about mapping camelCase to snake_case:

```python
@output_type
class Foo:
    first_arg: str
    second_arg: float
```

## Footnotes

1. RE: `input_property()` (and `output_property()`)

   - **Naming:** I very much wanted to collapse these two functions into a single `property` function, but doing so conflicts with Python's built-in `property` decorator. Should we just use `property` anyway and force fully qualifying `pulumi.property("bar")`? Or use something shorter like `prop("bar")` or `member("bar")`? Another idea is `field("bar")`, but I'd prefer to stay away from that as it may be confused with `dataclasses.field`. We could consider using `input("bar")`, which is nice, but I was concerned folks may confuse this with the `Input` class (and similarly for `output("bar")` and `Output`).

   - **Why not a decorator?** Decorators can only be applied to functions or classes.

   - **Why do we even need these?** Our generated code already uses mapping tables to map between snake_case and camelCase, but there are [issues with the current implementation](https://github.com/pulumi/pulumi/issues/3151). By explicitly specifying the names here, we can bypass the mapping process, and ensure the names are correct for these objects. This also paves the way to eventually addressing https://github.com/pulumi/pulumi/issues/3151 in a major version bump. These are analogous to the `InputAttribute` and `OutputAttribute` [used in .NET](https://github.com/pulumi/pulumi-aws/blob/18293bd434b6763cff4a66dd5f43419e0dd5a5df/sdk/dotnet/S3/Inputs/BucketWebsiteArgs.cs#L18).

   - **Alternatives?** Alternatively, the names could be specified explicitly in some other manner:

        Such as via an argument to the `output_type()` decorator:

        ```python
        @input_type(mapping={"first_arg": "firstArg", "second_arg": "secondArg"})
        class FooArgs:
            first_arg: Input[str]
            second_arg: Input[float]
        ```

        Or, via some special method (similar to the [`translate_input_property`/`translate_output_property` methods on `Resource`](https://github.com/pulumi/pulumi/blob/8e6b8fc79df9b7042e7e21eb4089f428fa6adc28/sdk/python/lib/pulumi/resource.py#L741-L761)):

        ```python
        @input_type
        class FooArgs:
            first_arg: Input[str]
            second_arg: Input[float]

            def _translate_property(self, prop: str) -> str:
                return {
                    "first_arg": "firstArg",
                    "second_arg": "secondArg",
                }.get(prop) or prop
        ```

        But both of these alternatives are more verbose to write.

**Comment by lblackstone** [comment]
I like your current proposal compared to the alternatives because it keeps changes localized to the fields rather than requiring changes to multiple locations in the code (e.g., if a new field is added).

 It would be nice to use the same function name for properties in either input/output definitions. Using the fully-qualified `pulumi.property` name seems reasonable to me.

**Comment by komalali** [comment]
I like what you've proposed because it addresses https://github.com/pulumi/pulumi/issues/4902

Re: naming - I agree with not overwriting the built-in `property` - your current proposal or `prop()` feel equally fine to me.

**Comment by pgavlin** [comment]
> Using the fully-qualified pulumi.property name seems reasonable to me.

+1 to this.

**Comment by justinvp** [comment]
I'm going with `pulumi.property`.

```python
@pulumi.output_type
class Foo(dict):
    first_value: str = pulumi.property("firstValue")
    """
    The first value's docstring.
    """
    second_value: float = pulumi.property("secondValue")
    """
    The second value's docstring.
    """
```

One issue with this syntax is that PyCharm and VS Code do not show docstrings in tooltips when hovering over members (e.g. when hovering over `first_value` in `res.foo.apply(lambda f: f.first_value)`).

(Python data classes have the same problem. The typical workaround is to specify a single top-level docstring for the entire class.)

Resources have had this same issue: docstrings aren't included in tooltips when hovering over outputs (e.g. hovering over `arn` in `my_bucket.arn`).

```python
class Bucket(pulumi.CustomResource):
    arn: pulumi.Output[str]
    """
    The ARN of the bucket. Will be of format `arn:aws:s3:::bucketname`.
    """
```

I suppose the reason these don't show up in tooltips is because Python doesn't really "use" docstrings declared after class annotations. The language allows them to be declared there (docstrings can be declared pretty much anywhere), but Python ignores them and does not expose any way to access them programmatically AFAICT (i.e. they aren't exposed as `__doc__` anywhere, like on classes and functions). They seem to be only used by doc generators like Sphinx. [PEP 224](https://www.python.org/dev/peps/pep-0224/) proposed formally adding these to the language, but was rejected.

## Design Tweak

If we used Python properties rather than class annotations, the docstrings *do* show up in tooltips in PyCharm and VS Code (though, it only works in VS Code when the new Pylance language service is installed & enabled, which is going to be the default lang service for Python in the future).

I'd like to propose the following design change to the input/output types:

We keep the current proposed annotation-based design as-is, as a short syntax for writing these by hand when you don't care about docstrings in tooltips on individual members, and add support for the following, which we'll use in our provider generated code:

```python
@pulumi.output_type
class Foo(dict):
    @builtins.property
    @pulumi.getter("firstValue")
    def first_value(self) -> str:
        """
        The first value's docstring.
        """
        ...

    @builtins.property
    @pulumi.getter("secondValue")
    def second_value(self) -> float:
        """
        The second value's docstring.
        """
        ...
```

The very first snippet I showed at the top of this comment that uses the shorter annotation-based syntax is equivalent to this expanded form (processed by the decorator; the docstrings on the annotations cannot be carried over, since they aren't accessible).

The `...` in the property function bodies is by-design. The ellipsis is similar to `pass` and is used to declare an empty function body. The `@pulumi.getter` decorator replaces empty functions with an implementation that looks up the value. This avoids having to specify the property name in multiple places (in both the `@pulumi.getter` decorator and function body to lookup the property value).

If you need to customize the implementation of the getter function, you can provide an implementation that calls `pulumi.get(self, "firstValue")`.

So this property with `...`:

```python
    @builtins.property
    @pulumi.getter("firstValue")
    def first_value(self) -> str:
        """
        The first value's docstring.
        """
        ...
```

is equivalent to:

```python
    @builtins.property
    @pulumi.getter("firstValue")
    def first_value(self) -> str:
        """
        The first value's docstring.
        """
        return pulumi.get(self, "firstValue")
```

Note that this fully-expanded form is similar to the generated JSII code for Python.

## Resources

I would like to address this for top-level Resource outputs (where it's arguably more important), but as a follow-up task. Here's my proposal for enabling it in a compatible way:

We add the following "protected" method to the base `Resource` class:

```python
class Resource:
    def _set_property(self, name: str, value: Any) -> None:
        self.__setattr__(name, value)
```

And update all the places where we set the output properties (e.g. in `transfer_properties` and where we set `id` and `urn`) to call `res._set_property(name, value)` rather than `res.__setattr__(name, value)`.

This way, the existing approach continues to work: by default, outputs will be set as instance variables on the resource.

We then update our provider codegen to emit the following:

```python
class Bucket(CustomResource):
    def __init__(self, ...):
        # ...
        self._values = dict()

    def _set_property(self, name: str, value: Any) -> None:
        self._values[name] = value
    
    @builtins.property
    @pulumi.getter("arn")
    def arn(self) -> pulumi.Output[str]:
        """
        The ARN of the bucket. Will be of format `arn:aws:s3:::bucketname`.
        """
        return self._values["arn"]
```

Our generated resources will store these in a private `_values` dict, and have Python properties for accessing each property, which will look up the value in `_values`.

This should be a compatible change, unless you're doing something out of the ordinary where it would matter that these are now properties rather than instance variables.

If we're concerned about compatibility implications, we may want to hold off on changing this for Resources until we're releasing major new versions of the providers. But I'd like to do it for input/output types now.

## Function Outputs

The class we generate for the result of calls to functions looks like:

```python
class GetBucketResult:
    """
    A collection of values returned by getBucket.
    """
    def __init__(__self__, ...):
        if arn and not isinstance(arn, str):
            raise TypeError("Expected argument 'arn' to be a str")
        __self__.arn = arn
        """
        The ARN of the bucket. Will be of format `arn:aws:s3:::bucketname`.
        """
        if bucket_domain_name and not isinstance(bucket_domain_name, str):
            raise TypeError("Expected argument 'bucket_domain_name' to be a str")
        __self__.bucket_domain_name = bucket_domain_name
        """
        The bucket domain name. Will be of format `bucketname.s3.amazonaws.com`.
        """
```

Here we declare a docstring after setting the instance variable (again, Python allows it, but ignores them). PyCharm will actually show these in tooltips, but VS Code does not. Presumably, we'd want to change these to be properties the same time we're converting the resources to use properties.

**Comment by komalali** [comment]
This looks good to me. Quite verbose but it'll have a much better end-user experience, and since we're autogenerating the code and maintaining the shorthand it sounds like a win-win to me.

**Comment by pgavlin** [comment]
I think that this looks good to me as well.

I'm wondering if we might be able to use this in the future to do destination-type-driven unmarshaling like we do in Go and .NET. The use of the attributes is pretty compelling.
